### ISSUE TITLE
Design choices for adding rerankers/cross-encoders  

### ISSUE THREAD

**Issue by orionw** [embeddings-benchmark/mteb#450]
@KennethEnevoldsen @Muennighoff 

Before I implement a solution, I wanted to sketch out my potential direction.

Cross-encoders need to take both query and document at the same time.  This is somewhat orthogonal to the early design of mteb (it's not about embeddings or the e in mteb) and so many abstract tasks will require some heavy changes as they expect embeddings (see [BitextMining](https://github.com/embeddings-benchmark/mteb/blob/main/mteb/evaluation/evaluators/BitextMiningEvaluator.py#L41) or [STS](https://github.com/embeddings-benchmark/mteb/blob/main/mteb/evaluation/evaluators/STSEvaluator.py#L44)).

I can see two options:

1. Update all abstract tasks to take cross-encoders, will require changing each abstract task to check for the model type and generate results from either of the options
2. Update just the most relevant tasks (Retrieval) and just don't allow rerankers in some of the other classes like BitextMining or STSEvaluator.

I would've liked to do this just by extending the DRESModel class without any other changes, but the AbstractTasks themselves make that impossible since Cross-Encoders can't return embeddings

I'm fine with either proposal, but wasn't sure how big of a change we wanted.  I'm also fine implementing it just for retrieval and instruction retrieval and then others can extend it later.  

--------

My thinking for how to structure this: 
- define a new model class `Reranker` (or `CrossEncoder`) that takes in the usual arguments like model name, batch size, etc. and  **also includes a DRESModel or a path to the results of a DRESModel**. The `reranker` class will implement a `search` function that only does reranking, taking in queries and documents at the same time based on the top N results from the DRESModel.
- Change the abstract tasks to check for this model type and if so, to pass both query/doc in rather than do the normal query and document separately.

Please let me know if there's another way I should be structuring this change!

**Comment by KennethEnevoldsen** [comment]
@orionw I have actually been considering adding in a model interface object (a python protocol). If you are familiar feel free to create a protocol for the CrossEncoder otherwise I will create one from the Encoder and then we can extend it from there.

**Comment by orionw** [comment]
> model interface object (a python protocol)

Do you mean like a pure abstract class / mixin style object?  I'm not very familiar with interfaces in Python.

Unfortunately, I don't think there is much similarity between CrossEncoders and the standard Encoders -- although the names suggest similarity, CrossEncoders don't really share any of the same methods.  CrossEncoders are really just another name for LLMs adapted to retrieval tasks these days.

> Is there any standard interfaces for CrossEncoders we can rely on for this?

A great question, also not really.  I was planning to make the `rerank` method throw an error if not extended as a type of abstract class to force users to define it.  There are a few general types of reranking, but usually it's model specific (a specific token true/false or a specific listwise reranking, etc.).

**Comment by Muennighoff** [comment]
Great discussion! I think that from a user-perspective, the best way would be to in addition to supporting

**Comment by Muennighoff** [comment]
Great discussion. I think that for users the best way would be to in addition to supporting
```python
from mteb import MTEB
from sentence_transformers import SentenceTransformer
model_name = "average_word_embeddings_komninos"
model = SentenceTransformer(model_name)
evaluation = MTEB(tasks=["Banking77Classification"])
results = evaluation.run(model, output_folder=f"results/{model_name}")
```
we also support
```python
from mteb import MTEB
from sentence_transformers import CrossEncoder
model_name = "average_word_embeddings_komninos"
model = CrossEncoder(model_name)
evaluation = MTEB(tasks=["Banking77Classification"])
results = evaluation.run(model, output_folder=f"results/{model_name}")
```
where CrossEncoder from SentenceTransformers could be the standard interface? Not sure how well it works but @tomaarsen may have thoughts.

Then users can combine them in anyway they want by passing previous result files and a value k for how much to rerank? I.e. you can also rerank with a bi-encoder if you want to.

Just my thoughts, I may be mistaken / other ways may be better! :)

**Comment by KennethEnevoldsen** [comment]
> Do you mean like a pure abstract class / mixin style object? I'm not very familiar with interfaces in Python.

A [protocol](https://peps.python.org/pep-0544/) is quite distinct from a abstract class in the case that it does not need to inherit e.g. a SentenceTransformer can implement a Encoder protocol (it is a class with an encode method) without ever inheriting from the Encoder protocol. You can see an example [here](https://github.com/KennethEnevoldsen/scandinavian-embedding-benchmark/blob/main/src/seb/interfaces/model.py#L22) in the SEB.

Basically it just specifies what methods you want from the object. A popular saying is: _"if it quacks [implements .quack()] like a duck and swim like a duck [has method .swim()] it is a duck [even though it isn't a subclass of duck]"_ ;)

It is not too different from an abstask really with the exception that it rarely defines functionality and does not require inheritance.

Python primatives like Sequence (e.g. list, tuples), Iterable,  Mapping (e.g dict)  are all protocols.

**Comment by KennethEnevoldsen** [comment]
Ahh right so maybe let us built the CrossEncoder interface around the sentence transformer interface if possible

**Comment by tomaarsen** [comment]
What a wonderful coincidence! Just yesterday I was drafting a potential extension of MTEB with CrossEncoder models! Or perhaps it's not a coincidence, if you were also inspired by yesterday's Jina reranker models.

In my experiments, I only attempted to extend MTEB for Retrieval, though I didn't get very far. MTEB has a sizable collection of high quality datasets, and (re-)using them for reranking seems quite valuable. That said, a big concern of mine was that CrossEncoder models are slow: they're not normally used for retrieval across a ton of documents, but instead for reranking. 

We might have to be a bit smart and selective with our datasets as a result, or select subsets instead, but I dont think this should be a problem.

Additionally, I think it makes sense to start small: we don't need to necessarily support all tasks out of gate, especially as CrossEncoders are primarily used for retrieval and reranking tasks.

Regarding using the SentenceTransformers `CrossEncoder` as a "standard" model: I think that would work quite well. Just like with the current bi-encoder implementation, we don't have to require that the model is a CrossEncoder instance, just that it supports the relevant methods: `__init__`, `predict` and `rank` (or variations, if desired, like `predict_batch`). @KennethEnevoldsen's idea regarding protocols works well with this, and CrossEncoders will fit the protocols nicely, allowing for easy tests as there's quite a few open reranker models that would benefit from benchmarking.

Additionally, I want to notify you that I intend to invest more time into improving CrossEncoders within Sentence Transformers. The bi-encoders are getting improved support through better training next month or so, and cross encoders are also in the pipeline afterwards. My ideas include easier training, loss logging, bfloat16 training, multi-gpu training, automatic model card generation, a "Use in Sentence Transformers" button on Hugging Face, etc.

I'd be glad to assist this effort, both here directly and via changes in Sentence Transformers, if beneficial. That also includes reviews, running benchmarks, etc. I think this can be very impactful for model authors, researchers and NLP practitioners alike.

- Tom Aarsen

**Comment by orionw** [comment]
Thanks for the comment @tomaarsen! That is wonderful news and I agree with your comments about them being very expensive. I did see the Jina rerankers but was doing this to incorporate some results from my recent [FollowIR paper](https://arxiv.org/abs/2403.15246), so unfortunately it was not as selfless as you are. Thanks for your help on SentenceTransformer and its really exciting to hear that CrossEncoders will be improved :)

I've made a PR that accomplishes this, those of you that are interested/able to review it please feel free to check out #457.

**Comment by KennethEnevoldsen** [comment]
I believe this issue can be closed - do re-open if there is something I missed
